File: cupom.py
from datetime import datetime

from bluetooth import BluetoothPrinter
from selector import selecionar_impressora
from utils import carregar_json, formatar_cpf, gerar_pix_copia_cola, remover_acentos


class MockPrinter:
    def __init__(self, colunas=32):
        self.linhas = []
        self.align = 'left'
        self.colunas = colunas
        self._current_line = ""

    def set(self, align='left', **kwargs):
        self.align = align

    def text(self, txt: str):
        partes = txt.split('\n')
        for i, parte in enumerate(partes):
            if i < len(partes) - 1:
                linha_completa = self._current_line + parte
                if self.align == 'center':
                    self.linhas.append(linha_completa.strip().center(self.colunas))
                elif self.align == 'right':
                    self.linhas.append(linha_completa.strip().rjust(self.colunas))
                else:
                    self.linhas.append(linha_completa)
                self._current_line = ""
            else:
                self._current_line += parte

    def cut(self):
        pass

    def qr(self, content, **kwargs):
        # Desenho ASCII simulando a machadada de um QR Code pro Preview
        self.linhas.append(" ")
        self.linhas.append("██████████████".center(self.colunas))
        self.linhas.append("██ SIMULA   ██".center(self.colunas))
        self.linhas.append("██ QR CODE  ██".center(self.colunas))
        self.linhas.append("██████████████".center(self.colunas))
        self.linhas.append(" ")

    def get_text(self):
        return "\n".join(self.linhas)


def _renderizar(p, loja: dict, pedido: dict, padding: int = 2):
    """Lógica central de impressão — agnóstica à fonte do printer."""

    nome_loja = loja.get('nome', 'ESTABELECIMENTO')

    colunas_suportadas = loja.get('largura_colunas', 30)
    util = colunas_suportadas - (padding * 2)
    pad  = " " * padding

    p.set(align='center')
    p.text(f"{nome_loja}\n\n")

    data_hora = datetime.now().strftime("%d/%m/%Y %H:%M")
    p.text(f"{data_hora}\n\n")

    cpf = pedido.get('cpf')
    if cpf:
        p.text(f"CPF: {formatar_cpf(cpf)}\n")

    p.set(align='left')
    p.text(f"{pad}{'-' * util}\n")

    l_qtd      = 5
    l_preco    = 8
    l_nome_max = util - l_qtd - l_preco

    p.text(f"{pad}{'QTD':<{l_qtd}}{'ITEM':<{l_nome_max}}{'TOTAL':>{l_preco}}\n")
    p.text(f"{pad}{'-' * util}\n")

    total_geral = 0

    for item in pedido.get('itens', []):
        nome_limpo = remover_acentos(str(item['nome']))
        qtd        = item.get('qtd')
        preco      = item.get('preco_uni', 0.0)

        if qtd is None:
            qtd_str    = ""
            total_item = preco
        else:
            qtd_str    = f"{qtd}x"
            total_item = qtd * preco

        total_geral += total_item

        preco_str    = f"{total_item:.2f}"
        nome_cortado = nome_limpo[:l_nome_max - 1]

        linha = f"{pad}{qtd_str:<{l_qtd}}{nome_cortado:<{l_nome_max}}{preco_str:>{l_preco}}\n"
        p.text(linha)

    p.text(f"{pad}{'-' * util}\n")

    descontos = pedido.get('descontos', [])
    if descontos:
        txt_sub = f"SUBTOT: R$ {total_geral:.2f}"
        p.text(f"{pad}{txt_sub.rjust(util)}\n")
        
        for idx, desc in enumerate(descontos):
            nome = remover_acentos(desc.get('nome') or f"Desc {idx+1}")
            tipo = desc.get('tipo', 'valor')
            valor = float(desc.get('valor', 0.0))
            
            if tipo == 'valor':
                abatido = valor
                str_desc = f"- R$ {valor:.2f}"
            else:
                abatido = (total_geral * valor) / 100.0
                str_desc = f"-{valor:g}% ({abatido:.2f})"
            
            total_geral -= abatido
            if total_geral < 0: total_geral = 0
            
            resto = util - len(str_desc)
            nome_c = nome[:resto - 1]
            p.text(f"{pad}{nome_c:<{resto}}{str_desc}\n")
            
        p.text(f"{pad}{'-' * util}\n")

    texto_final = f"TOTAL A PAGAR: R$ {total_geral:.2f}" if descontos else f"TOTAL: R$ {total_geral:.2f}"
    p.text(f"{pad}{texto_final.rjust(util)}\n")

    p.set(align='center')
    p.text("\nObrigado pela preferencia!\n\n")

    chave_pix   = loja.get('chave_pix')
    cidade_loja = loja.get('cidade', '')

    if chave_pix and total_geral > 0:
        pix_copia_cola = gerar_pix_copia_cola(chave_pix, total_geral, nome_loja, cidade_loja)
        p.text("Pague via PIX:\n\n")
        p.qr(pix_copia_cola, size=6, center=True)

    p.cut()

    if isinstance(p, BluetoothPrinter):
        p.flush()


def imprimir_cupom(padding: int = 2):
    """Fluxo interativo via terminal (seleciona impressora pelo menu)."""
    try:
        loja   = carregar_json("loja.json")
        pedido = carregar_json("pedido.json")
        p      = selecionar_impressora()
        print("Imprimindo cupom...\n")
        _renderizar(p, loja, pedido, padding)
        print("✅ Cupom impresso!")
    except ConnectionError as e:
        print(f"\n❌ Erro de conexão: {e}")
    except Exception as e:
        print(f"\n❌ Erro: {e}")


def imprimir_cupom_direto(printer, pedido: dict, padding: int = 2):
    """Chamada pelo servidor web — recebe printer e pedido já montados."""
    loja = carregar_json("loja.json")
    _renderizar(printer, loja, pedido, padding)

File: src/services/cupom.ts
/**
 * cupom.ts — port of cupom.py
 * Contains the full receipt layout logic and the MockPrinter for live preview.
 */

import { encodeText } from './escpos';
import { gerarPixCopiaECola } from './pix';
import { processLogo } from './image';
import type { Loja, Pedido } from '../types';

// ── Printer interface ─────────────────────────────────────────────────────────

/** Minimal interface satisfied by both BluetoothPrinter and MockPrinter. */
export interface Printer {
  set(align: 'left' | 'center' | 'right'): void;
  text(txt: string): void;
  image(base64: string): void;
  qr(content: string, size?: number): void;
  cut(): void;
}

// ── MockPrinter (live preview) ────────────────────────────────────────────────

/**
 * Implements Printer but renders to a text string instead of bytes.
 * Direct port of the Python MockPrinter class.
 */
export class MockPrinter implements Printer {
  private lines: string[] = [];
  private currentLine = '';
  private alignMode: 'left' | 'center' | 'right' = 'left';
  readonly colunas: number;

  constructor(colunas = 32) {
    this.colunas = colunas;
  }

  set(align: 'left' | 'center' | 'right' = 'left'): void {
    this.alignMode = align;
  }

  text(txt: string): void {
    const parts = txt.split('\n');
    for (let i = 0; i < parts.length; i++) {
      if (i < parts.length - 1) {
        const full = this.currentLine + parts[i];
        if (this.alignMode !== 'left') {
          this.lines.push(`[ALIGN:${this.alignMode}]${full}[/ALIGN]`);
        } else {
          this.lines.push(full);
        }
        this.currentLine = '';
      } else {
        this.currentLine += parts[i];
      }
    }
  }

  image(base64: string): void {
    this.lines.push(`[IMG:${base64}]`);
  }

  qr(_content: string, _size = 6): void {
    this.lines.push(' ');
    this.lines.push('██████████████'.padStart((this.colunas + 14) / 2).padEnd(this.colunas));
    this.lines.push('██ SIMULA   ██'.padStart((this.colunas + 14) / 2).padEnd(this.colunas));
    this.lines.push('██ QR CODE  ██'.padStart((this.colunas + 14) / 2).padEnd(this.colunas));
    this.lines.push('██████████████'.padStart((this.colunas + 14) / 2).padEnd(this.colunas));
    this.lines.push(' ');
  }

  cut(): void { /* no-op */ }

  getText(): string {
    return this.lines.join('\n');
  }
}

// ── Format helpers ────────────────────────────────────────────────────────────

function formatarCpf(cpf: string): string {
  if (cpf && cpf.length === 11) {
    return `${cpf.slice(0, 3)}.${cpf.slice(3, 6)}.${cpf.slice(6, 9)}-${cpf.slice(9)}`;
  }
  return cpf;
}

function removerAcentos(texto: string): string {
  return texto.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

// ── Core renderer ─────────────────────────────────────────────────────────────

/**
 * Central print logic — printer-agnostic.
 * Direct port of _renderizar() from cupom.py.
 */
export function renderizarCupom(
  p: Printer,
  pedido: Pedido,
  loja: Loja,
  padding = 2,
): void {
  const nomeLoja = loja.nome || 'ESTABELECIMENTO';
  const colunas  = loja.largura_colunas || 32;
  const util     = colunas - padding * 2;
  const pad      = ' '.repeat(padding);

  p.set('center');
  if (loja.logo) {
    p.image(loja.logo);
    p.text('\n');
  }
  p.text(`${nomeLoja}\n\n`);

  const now = new Date();
  const dataHora = `${now.toLocaleDateString('pt-BR')} ${now.toLocaleTimeString('pt-BR', { hour: '2-digit', minute: '2-digit' })}`;
  p.text(`${dataHora}\n\n`);

  if (pedido.cpf) {
    p.text(`CPF: ${formatarCpf(pedido.cpf)}\n`);
  }

  p.set('left');
  p.text(`${pad}${'-'.repeat(util)}\n`);

  const lQtd     = 5;
  const lPreco   = 8;
  const lNomeMax = util - lQtd - lPreco;

  p.text(`${pad}${'QTD'.padEnd(lQtd)}${'ITEM'.padEnd(lNomeMax)}${'TOTAL'.padStart(lPreco)}\n`);
  p.text(`${pad}${'-'.repeat(util)}\n`);

  let totalGeral = 0;

  for (const item of pedido.itens) {
    const nomeLimpo = removerAcentos(String(item.nome));
    const qtd       = item.qtd;
    const preco     = item.preco_uni ?? 0;

    const qtdStr    = qtd === null ? '' : `${qtd}x`;
    const totalItem = qtd === null ? preco : qtd * preco;
    totalGeral      += totalItem;

    const precoStr    = totalItem.toFixed(2);
    const nomeCortado = nomeLimpo.slice(0, lNomeMax - 1);

    const linha = `${pad}${qtdStr.padEnd(lQtd)}${nomeCortado.padEnd(lNomeMax)}${precoStr.padStart(lPreco)}\n`;
    p.text(linha);
  }

  p.text(`${pad}${'-'.repeat(util)}\n`);

  const descontos = pedido.descontos ?? [];
  if (descontos.length > 0) {
    const txtSub = `SUBTOT: R$ ${totalGeral.toFixed(2)}`;
    p.text(`${pad}${txtSub.padStart(util)}\n`);

    for (let idx = 0; idx < descontos.length; idx++) {
      const desc  = descontos[idx];
      const nome  = removerAcentos(desc.nome || `Desc ${idx + 1}`);
      const tipo  = desc.tipo;
      const valor = Number(desc.valor);

      let abatido: number;
      let strDesc: string;

      if (tipo === 'valor') {
        abatido = valor;
        strDesc = `- R$ ${valor.toFixed(2)}`;
      } else {
        abatido = (totalGeral * valor) / 100;
        strDesc = `-${valor}% (${abatido.toFixed(2)})`;
      }

      totalGeral -= abatido;
      if (totalGeral < 0) totalGeral = 0;

      const resto  = util - strDesc.length;
      const nomeC  = nome.slice(0, resto - 1);
      p.text(`${pad}${nomeC.padEnd(resto)}${strDesc}\n`);
    }

    p.text(`${pad}${'-'.repeat(util)}\n`);
  }

  const textoFinal =
    descontos.length > 0
      ? `TOTAL A PAGAR: R$ ${totalGeral.toFixed(2)}`
      : `TOTAL: R$ ${totalGeral.toFixed(2)}`;
  p.text(`${pad}${textoFinal.padStart(util)}\n`);

  p.set('center');
  if (loja.mensagem_rodape?.trim()) {
    p.text(`\n${loja.mensagem_rodape}\n\n`);
  } else {
    p.text('\n');
  }

  const chavePix   = loja.chave_pix;
  const cidadeLoja = loja.cidade ?? '';

  if (chavePix && totalGeral > 0) {
    const pixStr = gerarPixCopiaECola(chavePix, totalGeral, nomeLoja, cidadeLoja);
    if (pixStr) {
      p.text('Pague via PIX:\n\n');
      p.qr(pixStr, 6);
    }
  }

  p.cut();
}

// ── Public helpers ────────────────────────────────────────────────────────────

/** Returns a live-preview text string without printing. */
export function gerarPreview(pedido: Pedido, loja: Loja, isPremium = false): string {
  const mock = new MockPrinter(loja.largura_colunas);
  
  // Create a temporary loja object with the processed logo for rendering
  const lojaParaRender = { ...loja };
  
  // SÓ PROCESSA A LOGO PARA O PREVIEW SE FOR PREMIUM
  if (loja.logo_raw && isPremium) {
    lojaParaRender.logo = processLogo(loja.logo_raw, loja.logo_metodo || 'dither');
  } else {
    delete lojaParaRender.logo;
  }

  renderizarCupom(mock, pedido, lojaParaRender);
  return mock.getText();
}
